pubsub/gcloud/aio/pubsub/subscriber.py

```diff
@@ -0,0 +1,222 @@
+
+        def cache_outdated(self) -> bool:
+            if (time.perf_counter() - self.last_refresh) > self.cache_timeout:
+                return True
+            return False
+
+    async def acker(subscription: str,
+                    ack_queue: 'asyncio.Queue[str]',
+                    subscriber_client: 'SubscriberClient',
+                    ack_window: float,
+                    metrics_client: MetricsAgent) -> None:
+        ack_ids: List[str] = []
+        while True:
+            time_budget = ack_window
+            if not ack_ids:
+                ack_ids.append(await ack_queue.get())
+                ack_queue.task_done()
+
+            while time_budget > 0:
+                start = time.perf_counter()
+                try:
+                    ack_id = await asyncio.wait_for(
+                        ack_queue.get(), timeout=time_budget)
+                    ack_ids.append(ack_id)
+                    ack_queue.task_done()
+                except asyncio.TimeoutError:
+                    break
+                time_budget -= (time.perf_counter() - start)
+
+            try:
+                await subscriber_client.acknowledge(subscription,
+                                                    ack_ids=ack_ids)
+            except Exception as e:
+                log.warning(
+                    'Ack request failed, better luck next batch', exc_info=e)
+                continue
+
+            metrics_client.histogram('pubsub.acker.batch', len(ack_ids))
+
+            ack_ids = []
+
+    async def _fire_callback(semaphore: asyncio.Semaphore,
+                             message: SubscriberMessage,
+                             callback: ApplicationHandler,
+                             ack_queue: 'asyncio.Queue[str]',
+                             metrics_client: MetricsAgent
+                             ) -> None:
+        try:
+            start = time.perf_counter()
+            await callback(message)
+            await ack_queue.put(message.ack_id)
+            metrics_client.increment('pubsub.consumer.succeeded')
+            metrics_client.histogram('pubsub.consumer.latency.runtime',
+                                     time.perf_counter() - start)
+        except Exception:
+            log.exception('Application callback raised an exception')
+            metrics_client.increment('pubsub.consumer.failed')
+        finally:
+            semaphore.release()
+
+    async def consumer(
+            message_queue: MessageQueue,
+            callback: ApplicationHandler,
+            ack_queue: 'asyncio.Queue[str]',
+            ack_deadline_cache: AckDeadlineCache,
+            consumer_pool_size: int,
+            metrics_client: MetricsAgent) -> None:
+        semaphore = asyncio.Semaphore(consumer_pool_size)
+        while True:
+            await semaphore.acquire()
+
+            message, pulled_at = await message_queue.get()
+
+            ack_deadline = await ack_deadline_cache.get()
```

Comment:
Thinking more about the ackDeadline stuff, it occurs to me that we're only doing this check upfront rather than immediately before acking, so we'd only catch a subset of the potential failures here (eg. imagine ackDeadline=5s, the message sits on the queue for 4s [so this check passes], then takes >1s to process). I wonder if this best-effort only-a-subset-of-cases failfast stuff is worth having at all or if its just a premature optimization. Perhaps we should remove this entirely?

IIUC a good chunk of the reason to have this is for metrics reporting, eg. so users know to tune their ackDeadline value. Is there another way we can get this? What happens if we `ack` a message which has exceeded it's deadline -- do we get a response there which we could use to get the metric?

Reply:
What you saying is true, failfast is by no means exhaustive. It will only make sure we don't run potentially heavy processing on lost cause messages and show it in metrics.

>do we get a response there which we could use to get the metric?

"Acknowledging a message whose ack deadline has expired may succeed, but such a message may be redelivered later."
I'm not 100% sure what they mean by _may succeed_, but in practice it always succeeds.

Reply:
> What you saying is true, failfast is by no means exhaustive. It will only make sure we don't run potentially heavy processing on lost cause messages and show it in metrics.

I think I'd be ok with that, assuming we can get the failfast metric elsewhere -- if a user has set their ackDeadline to `n` seconds and it takes `>n` seconds to process them, they'll have to tweak settings anyway. If we can make our system simpler and easier to reason about in exchange for that case also burning CPU cycles, that seems like a fair tradeoff to me (note that we're not actually wasting processing time in that case anyway, we're just replacing noop loops with calls to their worker; they'd have to be running the app already anyway so nothing is really "wasted" here)

> I'm not 100% sure what they mean by may succeed, but in practice it always succeeds.

Well that's garbage. Maybe this should be moved to the acker thread then? It's worth comparing against the ackdeadline to get that failfast metric to help users debug, but it would be even more valuable if they could rely on it as a 100% source of truth.

Reply:
Not all CPU cycles are equal, we'll be running blocking single-threaded models in that callback, won't we? Every second of processing there adds +1 to each pending message timer. Not to say that 1 second could be spent on a message that could actually be acked. So they're "wasted" in economics sense :D
Moving this metric to acker will take away all the value from that `if` in exchange for a metric that is not 100% accurate, because there is a time gap between acker reporting metric and making a request.

By the way, google now exposes `subscription/ack_latencies` metric so we already have a metric for the whole chain.

Reply:
> By the way, google now exposes `subscription/ack_latencies` metric so we already have a metric for the whole chain.

Oh, fantastic, that renders this whole conversation moot then -- happy to rely on their metrics for this rather than trying to coerce our own system into giving us a copy of that data. Sounds good :+1: